src/client.js

```diff
@@ -306,34 +311,54 @@ module.exports = class IrcClient extends EventEmitter {
     }
 
     startPeriodicPing() {
-        const that = this;
-        let timeout_timer = null;
+        const client = this;
+        let ping_timer = null;
 
-        if (that.options.ping_interval <= 0 || that.options.ping_timeout <= 0) {
+        if (client.options.ping_interval <= 0) {
             return;
         }
 
         // Constantly ping the server for lag and time syncing functions
         function pingServer() {
-            that.ping();
-            that.connection.setTimeout(pingServer, that.options.ping_interval * 1000);
+            client.ping();
+        }
+
+        function resetPingTimer() {
+            client.connection.clearTimeout(ping_timer);
+            ping_timer = client.connection.setTimeout(pingServer, client.options.ping_interval * 1000);
+        }
+
+        // Browsers have started throttling looped timeout callbacks
+        // using the pong event to set the next ping breaks this loop
+        this.command_handler.on('pong', resetPingTimer);
+
+        // Start timer
+        resetPingTimer();
+    }
+
+    startPingTimeoutTimer() {
+        const client = this;
+        let timeout_timer = null;
+
+        if (client.options.ping_timeout <= 0) {
+            return;
         }
 
         // Data from the server was detected so restart the timeout
         function resetPingTimeoutTimer() {
-            that.connection.clearTimeout(timeout_timer);
-            timeout_timer = that.connection.setTimeout(pingTimeout, that.options.ping_timeout * 1000);
+            client.connection.clearTimeout(timeout_timer);
+            timeout_timer = client.connection.setTimeout(pingTimeout, client.options.ping_timeout * 1000);
         }
 
         function pingTimeout() {
-            that.debugOut('Ping timeout (' + that.options.ping_timeout + ' seconds)');
-            that.emit('ping timeout');
-            const end_msg = that.rawString('QUIT', 'Ping timeout (' + that.options.ping_timeout + ' seconds)');
-            that.connection.end(end_msg, true);
+            client.debugOut('Ping timeout (' + client.options.ping_timeout + ' seconds)');
+            client.emit('ping timeout');
+            const end_msg = client.rawString('QUIT', 'Ping timeout (' + client.options.ping_timeout + ' seconds)');
+            client.connection.end(end_msg, true);
         }
 
         this.resetPingTimeoutTimer = resetPingTimeoutTimer;
```

Comment:
Do you think this whole thing could be simplified to avoid assigning functions here?

Reply:
Not sure, I'im assuming it was done like this originally to keep all the related code together and to ensure the correct `this` is used when being called from other parts of the code and potentially setTimeout's